[macOS] Safari sometimes stalls when entering window fullscreen from Start Page

whsieh opened the pull request:
#### 807c526fad673c68ec2691e3e31657caa8e17f94
<pre>
[macOS] Safari sometimes stalls when entering window fullscreen from Start Page
<a href="https://bugs.webkit.org/show_bug.cgi?id=295931">https://bugs.webkit.org/show_bug.cgi?id=295931</a>
<a href="https://rdar.apple.com/152090882">rdar://152090882</a>

Reviewed by Tim Horton.

This is a speculative fix for <a href="https://rdar.apple.com/152090882">rdar://152090882</a>, wherein Safari becomes (visually) unresponsive when
entering window fullscreen, due to the ObjC block returned by `-_holdResizeSnapshotWithReason:`
staying around for much longer than intended. I wasn&apos;t able to reproduce this locally, but we can
implement several mitigations to improve this:

1.  Avoid holding the resize snapshot in the case where the web process is already known to be
    unresponsive, and is unlikely to complete a presentation update in a reasonable amount of time
    anyways.

2.  Avoid holding the resize snapshot in the case where the web process has been suspended, for the
    same reason.

3.  Avoid holding the resize snapshot if the web view `-isHidden`. There&apos;s no point in blocking on
    web content painting, since the page won&apos;t be visible to the user anyways.

4.  If all else fails, install a watchdog timer to ensure that the resize snapshot is invalidated
    after a fairly short delay (1 second), which should be more than sufficient for a single
    presentation update. This is currently unbounded, leading to the whole Safari app being
    unresponsive for an indefinite (in the bug, anywhere from 5 seconds to minute-long) amount of
    time.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _invalidateWindowSnapshotReadinessHandler]):
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _holdWindowResizeSnapshotIfNeeded]):

See above for more details. Additionally add more release logging, to make it easier to root-cause
similar issues with resize snapshots moving forward.

(-[WKWebView _doWindowSnapshotReadinessUpdate]):

Make this invoke `-_invalidateWindowSnapshotReadinessHandler` after 1 second, using
`-performSelector:withObject:afterDelay:`. This operation is canceled early if the presentation
update finishes first.

Canonical link: <a href="https://commits.webkit.org/297374@main">https://commits.webkit.org/297374@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b7fbef6573106b7ba1b1b7c7abd21dd909ed258b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111552 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31219 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21683 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117588 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/61823 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31899 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39800 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84764 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/61823 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114499 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25459 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100397 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65205 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24812 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61415 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94855 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18597 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120822 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38605 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28685 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93674 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38982 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96664 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93497 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23829 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38622 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16407 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34641 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38490 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43967 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38157 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41493 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39859 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b7fbef6573106b7ba1b1b7c7abd21dd909ed258b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48010

whsieh commented:
Thanks for the review!

whsieh committed:
[macOS] Safari sometimes stalls when entering window fullscreen from Start Page
https://bugs.webkit.org/show_bug.cgi?id=295931
rdar://152090882

Reviewed by Tim Horton.

This is a speculative fix for rdar://152090882, wherein Safari becomes (visually) unresponsive when
entering window fullscreen, due to the ObjC block returned by `-_holdResizeSnapshotWithReason:`
staying around for much longer than intended. I wasn't able to reproduce this locally, but we can
implement several mitigations to improve this:

1.  Avoid holding the resize snapshot in the case where the web process is already known to be
    unresponsive, and is unlikely to complete a presentation update in a reasonable amount of time
    anyways.

2.  Avoid holding the resize snapshot in the case where the web process has been suspended, for the
    same reason.

3.  Avoid holding the resize snapshot if the web view `-isHidden`. There's no point in blocking on
    web content painting, since the page won't be visible to the user anyways.

4.  If all else fails, install a watchdog timer to ensure that the resize snapshot is invalidated
    after a fairly short delay (1 second), which should be more than sufficient for a single
    presentation update. This is currently unbounded, leading to the whole Safari app being
    unresponsive for an indefinite (in the bug, anywhere from 5 seconds to minute-long) amount of
    time.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _invalidateWindowSnapshotReadinessHandler]):
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _holdWindowResizeSnapshotIfNeeded]):

See above for more details. Additionally add more release logging, to make it easier to root-cause
similar issues with resize snapshots moving forward.

(-[WKWebView _doWindowSnapshotReadinessUpdate]):

Make this invoke `-_invalidateWindowSnapshotReadinessHandler` after 1 second, using
`-performSelector:withObject:afterDelay:`. This operation is canceled early if the presentation
update finishes first.

Canonical link: https://commits.webkit.org/297374@main
